scripts/Test-Signatures.ps1

```diff
@@ -0,0 +1,297 @@
+        Write-Host
+    }
+}
+
+function Undo-Operation($Operation, $Message, [switch]$NoNewLine, [switch]$Suppress) {
+    if($Suppress) {
+        return
+    }
+    
+    Clear-Line
+    Write-Host "$('  ' * $Depth)[$Operation] " -ForegroundColor Red -NoNewline
+    Write-Host $Message -NoNewline
+
+    if (-not $NoNewLine -or $CIBuild) {
+        Write-Host
+    }
+}
+
+Function Invoke-Command ($Command, $Arguments)
+{
+    $pinfo = New-Object System.Diagnostics.ProcessStartInfo
+    $pinfo.FileName = $Command
+    $pinfo.RedirectStandardError = $true
+    $pinfo.RedirectStandardOutput = $true
+    $pinfo.UseShellExecute = $false
+    $pinfo.Arguments = $Arguments
+    $p = New-Object System.Diagnostics.Process
+    $p.StartInfo = $pinfo
+    $p.Start() | Out-Null
+    $p.WaitForExit()
+    
+    [pscustomobject]@{
+        stdout = $p.StandardOutput.ReadToEnd()
+        stderr = $p.StandardError.ReadToEnd()
+        ExitCode = $p.ExitCode
+    }
+}
+
+function Test-NuGetSignatures {
+    $failed = @{}
+    
+    $nupgks = Get-ChildItem *.nupkg
+ 
+    $nugetDir = "$PSScriptRoot\..\tools\nuget"
+    $nugetPath = "$nugetDir\nuget.exe"
+    
+    Push-Depth
+    if(-not (Test-Path $nugetDir))
+    {
+        # Create the directory for nuget.exe if it does not exist
+        New-Item -ItemType Directory -Force -Path $nugetDir | Out-Null
+
+        try {
+            $progressPreference = 'silentlyContinue'
+            $nugetUrl = "https://dist.nuget.org/win-x86-commandline/v4.6.1/nuget.exe"
+            Update-Operation -operation "downloading" -message $nugetUrl -suppress:$CIBuild
+            Invoke-WebRequest $nugetUrl -OutFile $nugetPath 
+            Submit-Operation -operation "downloaded" -message $nugetUrl
+        }
+        finally {
+            $progressPreference = 'Continue'
+        }
+    }
+
+    $nupgks | ForEach-Object {
+        $relativePath = $_.Name
+        Update-Operation -operation "processing" -message $relativePath -suppress:$CIBuild
+
+        $verificationProcess = Invoke-Command $nugetPath ("verify","-signature","-CertificateFingerprint","3F9001EA83C560D712C24CF213C3D312CB3BFF51EE89435D3430BD06B5D0EECE", $_.FullName)
+        if ($verificationProcess.ExitCode -eq 0) {
+            Submit-Operation -operation "valid" -message $relativePath
+        }
+        else {
+            Undo-Operation -operation "unsigned" -message $relativePath -NewLine
+            $failed.Add($_.FullName, $null)
+        }
+    }
+
+    if($failed.Count -eq 0) {
+        Submit-Operation -operation "done" -message "$($nupgks.Length) packages passed signature checks."
+    } elseif ($failed.Count -eq $nupgks.Length) {
+        Undo-Operation -operation "failed" -message "$($nupgks.Length) packages failed signature checks." 
+    } else {
+        Undo-Operation -operation "partial" -message "$($nupgks.Length - $failed.Count) packages passed signature checks and $($failed.Count) packages failed!" 

```

Comment:
I don't get why this distinction is necessary, seeing that 0 packages passed is good.